core/src/main/java/com/oath/oak/OakRValueBufferImpl.java

```diff
@@ -134,31 +166,47 @@ public double getDouble(int index) {
         if (transformer == null) {
             throw new NullPointerException();
         }
-        T retVal = ValueUtils.transform(bb, transformer);
-        if (retVal == null) {
+        Result<T> result = operator.transform(getValueSlice(), transformer, version);
+        if (result.operationResult == FALSE) {
             throw new ConcurrentModificationException();
+        } else if (result.operationResult == RETRY) {
+            lookupValueReference();
+            transform(transformer);
         }
-        return retVal;
+        return result.value;
     }
 
     @Override
     public void unsafeCopyBufferToIntArray(int srcPosition, int[] dstArray, int countInts) {
-        start();
-        ByteBuffer dup = ValueUtils.getValueByteBufferNoHeader(bb);
-        ValueUtils.unsafeBufferToIntArrayCopy(dup, srcPosition, dstArray, countInts);
-        end();
+        Slice s = getValueSlice();
+        start(s);
+        ByteBuffer dup = operator.getValueByteBufferNoHeader(s);
+        operator.unsafeBufferToIntArrayCopy(dup, srcPosition, dstArray, countInts);
+        end(s);
     }
 
-    private void start() {
-        // if the value was moved a ConcurrentModificationException is thrown
-        ValueUtils.ValueResult res = ValueUtils.lockRead(bb);
-        if (res != ValueUtils.ValueResult.SUCCESS) {
+    private void start(Slice valueSlice) {
+        ValueUtils.ValueResult res = operator.lockRead(valueSlice, version);
+        if (res == FALSE) {
```

Comment:
How lock can return false? If value was deleted? Please comment

Reply:
I added comments to the API of ValueUtils (in this case, the documentation of lockRead explains in what case FALSE can be returned).